README.md
# ESP32 RMT RC PWM Library
## Introduction
**Warning** <br>
This collection of cpp functions is the result of my first involvement with the ESP32 and especially its RMT hardware. Even though it works for me since a while without any problems, But i would not call it a robust plug&play Arduino-ESP32-lib.
It is meant for using the ESP32 with the Arduino Framework **Platform: espressif32@6.1.0** (compatible with older framework espressif32@3.2.0) and was developed with the [Platformio](https://platformio.org/install/ide?install=vscode) extension in [VSCode](https://code.visualstudio.com).
It uses the special hardware component RMT, so it can only be used for ESP32 chips with this component.

### Why did I create this library?

This library was specifically developed for working with older RC receivers, such as the Robbe Luna in my case. These receivers have some unpleasant characteristics compared to modern digital types, which can make it challenging to use them as input devices for microcontrollers. Specifically, these characteristics include:

1. Almost random and chaotic output signals when no transmitter is active. This can be extremely dangerous, especially when controlling motors. (see picture in examples)
2. The PWM frequency may not necessarily comply with the current standards or the technical specifications mentioned in the receiver's manual. This discrepancy can be attributed to aging discrete components that are decades old in some cases.
3. The range of pulse widths (min/max/neutral) can also deviate significantly from the expected values.

The purpose of this library is to address and handle these issues.

- To tackle point 1, a state machine has been implemented to detect whether a signal is stable. The states are categorized as `NO_SIGNAL`, `UNSTABLE`, and `STABLE`. The failsafe function returns `pulse_neutral`, similar to what is commonly done with newer digital receivers. However, unlike those receivers, the library provides the ability to determine whether the transmitter is still active by checking the status.

This is a CPP library for the ESP32-S1 microcontroller that utilizes the RMT hardware to handle RC PWM signals. The library provides the following features:

- Utilization of the RMT hardware to offload the processor.
- Simultaneous reading of up to 8 RC PWM channels.
- Support for reading both raw PWM values (`pwm_get_rawPwm`) and scaled values (`pwm_get_scaledPwm`).
- Scaling of PWM pulse based on predefined constants (or automatic measurement of min/max values), allowing individual adjustment of minimum, neutral and maximum values. (Default: min=1000, neutral=1500, max=2000  all in µs) 
- Separate scaling of the min-neutral and neutral-max ranges. For example, if the min/max calibration defines a min of 800, max of 2100, and neutral of 1200, the range 800-1200 will be scaled to 1000-1500, and 1200-2100 will be scaled to 1500-2000.
- Support for different PWM frequencies per channel (tested range: 20Hz - 200Hz). The phase relationship between channels is irrelevant. Compatibility with various receivers that may be connected to the channels. To avoid uint32_t overflow for lower frequencies, a prescaler must be set.

The library provides the following readings for each channel:

- Pulse width in µs.
- PWM frequency in Hertz (can be different per channel).
- Duty cycle as a percentage based on the actual min/max values.
- Offset in µs, representing the phase shift between the current channel and channel 0. This value is only meaningful if channel 0 and the respective channel operate synchronously (using the same receiver), which is not always the case. Otherwise, 0 is returned.

Each channel has a status associated with it, which is also provided:

- `NO_SIGNAL`: No signal detected on the channel.
- `UNSTABLE`: Unstable signal. For example, when using older RC receivers with no active signal from the transmitter. To be considered stable, the pulse width must fall within the valid range (configurable through constants), and the pulse width of two consecutive pulses must not differ by more than +/-20%. Additionally, the last pulse must not be older than 2 times `period_max` (configurable constant). Only when all conditions are met for multiple consecutive pulses, the signal transitions from UNSTABLE to STABLE. The required number of consecutive stable pulses can be configured through a constant.
- `STABLE`: Standard OK state.

The status of each channel can be manually reset to initiate a new calibration or perform other actions.

The library also supports automatic detection of the neutral position, where the current position is stored as the neutral position upon startup. This behavior can be enabled or disabled through software.

Automatic calibration of the min/max values is supported. During the first 5 seconds after startup, moving the control stick (or equivalent) to the extreme positions will save those values as the min/max values. The calculation of the duty cycle is then based on these values. The duration and the possibility of calibration can be configured through software. Optionally, during calibration, the current PWM value or `pulse_neutral` can be returned.

A failsafe function is implemented, which returns `pulse_neutral` for channels that do not have the STABLE status. However, the status check still allows determining whether the respective channel is receiving a valid signal. This behavior can be configured through a constant.

## How it works
You can find a more detailed description here
[HowItWorks](docs/HowItWorks.md)

## Installation
At this stage it is not a released Arduino Lib. Therefore you have to download the lib files manually and include them in your project.<br>

1. Download the lib-files <br>

lib/esp32-rmt-pwm-reader/esp32-rmt-pwm-reader.h <br>
lib/esp32-rmt-pwm-reader/esp32-rmt-pwm-reader.cpp <br>

2. Copy the files into a lib directory of your project.<br> 
If you are working with PlatformIO, you can use the directory structure of this repository.

## Usage
An example implementation can be found in esp32_pwm_read.cpp. Only the basic steps are shown here.

### Include the Library
```cpp
#include <esp32-rmt-pwm-reader.h>
// #include "esp32-rmt-pwm-reader.h" // if the lib is located directly in the project directory
````

### Initialize
Set which pin to use, configure the default values (if necessary) and start the esp32 rmt-read functionality

```cpp
// init channels and pins
uint8_t pins[] = {13, 14, 33, 27, 32};  // desired input pins
int numberOfChannels = sizeof(pins) / sizeof(uint8_t);

void setup() {
    // init channels
    pwm_reader_init(pins, numberOfChannels);

    // here you can change channel defaults values before reading (if needed)
    // e.g. pwm_set_channel_pulse_min() /max/neutral
    // e.g. set auto_zero/auto_min_max for channel 0-2
    for (int ch = 0; ch < 3; ch++) {
        pwm_set_auto_zero(ch, true);     // set channel to auto zero
        pwm_set_auto_min_max(ch, true);  // set channel to auto min/max calibration
    }

    // begin reading 
    esp_err_t err = pwm_reader_begin();
    if (err != ESP_OK) {
        Serial.printf("begin() err: %i", err);
    }
}

```
### Read the data of interest 
The lib offers many redundant functions for reading the individual data but also access to the complete data structure. Details can be found in the function descriptions in the [esp32-rmt-pwm-reader.h](lib/esp32-rmt-pwm-reader/esp32-rmt-pwm-reader.h).
In [esp_pwm_read.cpp](src/esp32_pwm_read.cpp) different variants are shown.  
```cpp
void loop(){
    // Reading the actual pulse width of channel 1 
    int32_t impulseChannel_2 = pwm_get_rawPwm(2); 

    // Do something with the pulse width...

    // The lib is not blocking. Therefore the read cycle can be 
    // quite high. In esp32_pwm_read.cpp I use 100ms. However, this 
    // should certainly not be necessary with a PWM frequency of 50Hz. 
    // delay() is just as an example. In fact you should use your own 
    // non-blocking task. e.g. like in esp32_pwm_read.cpp
    delay(200); 
}
```

### cleanup
In the vast majority of applications the pwm-reader will be active during the complete application. In this case, a cleanup is not necessary.<br>
However, it may be that the reader is only a partial functionality that should be activated only when needed. This is the case, for example, with ServoTester deluxe.<br>
In this case the Lib must provide for a tidy cleanup, so that the necessary Ressorcen are released. The pwm_cleanup() function does the job.<br>
In this scenario, the initialization cannot simply be done in the setup function, since that is called only once at program startup.
Instead, starting and stopping the PWM reader should be outsourced to separate functions.
In [ServoTester deluxe](https://github.com/rewegit/Servotester_Deluxe) you can see a simple variant for this.

## Examples
The sample application simply outputs a table of the most important values of all channels.
In addition, the most important access types to data and config are shown in the source code.
The table itself is output in the terminal using escape sequences. It is assumed that the default tab width of the shell used is 8 characters. 
If the output is messy, reducing the output speed can help.

```cpp
setup() {
    ...
    // decrase output to every 200ms
    ticker.attach_ms(200, readPwmSignals);  

}
```

Tested on MacOS with VSCode. The integrated shell was ZSH.


![unstable:](Images/ESP32_RMT_Example_Robbe_Luna_UNSTABLE.gif) 
Transmitter switched off<br>

![Receiver switched on](<./Images/ESP32_RMT_Example_Robbe_Luna_STABLE.gif>)
Transmitter switched on. control stick channel 1 is moved<br><br>

The following data are output:

| output | description |
| ------ | ----------- |
| Channel| channel number 
| offset | distance of the rising edge of the channel to the rising edge of channel 0 in µs
| raw    | currently read pwm value
| scaled | current pwm value scaled to min/neutral/max
| period | period of the channel in µs
| freq Hz | pwm frequency of the channel
| p_min | minimum pwm value preset or determined
| p_ntrl | neutral pwm value preset or determined
| p_max | maximum pwm value preset or determined
| dc | duty cycle calculated with raw and period
| state | current channel status 

Another example shows the integration of the lib in my fork of the [Servotester Deluxe](https://github.com/rewegit/Servotester_Deluxe). 

## History
see [Changelog](docs/Changelog.md)

## License
[MIT License](https://github.com/kapraran/FreqCountESP/blob/master/LICENSE)

## Links
 * [Espressif IoT Development Framework for ESP32](https://github.com/espressif/esp-idf)

## Acknowledgements
For inspiration <br>
 * [Arduino RC engine sound & light controller for ESP32](https://github.com/TheDIYGuy999/Rc_Engine_Sound_ESP32)
 * [Servotester_Deluxe](https://github.com/Ziege-One/Servotester_Deluxe)
 * [Servotester_Deluxe (with improved software)](https://github.com/TheDIYGuy999/Servotester_Deluxe)


docs/Changelog.md
# Changelog

## 0.2:
- its now compatible with older framework espressif32@3.2.0 

lib/esp32-rmt-pwm-reader/esp32-rmt-pwm-reader.h
/**
*********************************************************************************
* @file:	esp32-rmt-pwm-reader.h
* @author:	R.W.
* @date:	28.04.23
* @brief:
* @history: 20.07.23 rw compatibility with older versions of espressif framework
*                       especially espressif32@3.2.0 for esp32 Soundcontroller
*********************************************************************************
**/

#ifndef esp32_rmt_pwm_reader_h
#define esp32_rmt_pwm_reader_h

#include <Arduino.h>
#include <esp_timer.h>       // for meassure time

#include "driver/rmt.h"      // for PWM signal detection
#include "soc/rmt_reg.h"     // for rmt channel status
#include "soc/rmt_struct.h"  // for rmt ir

#include "driver/gpio.h"  // for direct gpio access

#ifndef GPIO_NUM_NC  // not defined in older versions of gpio.h
#define GPIO_NUM_NC -1

#endif  //GPIO_NUM_NC

// #define PIN_TST GPIO_NUM_4


#include "esp32-hal-log.h"  // loglevel ist set as build_flags = -DCORE_DEBUG_LEVEL in platformio.ini
static const char *TAG = "rmt_pwm_reader";

// defines for pulse calculations
#define STABLE_COUNTER 10  // periods that must be in limit before status changes from UNSTABLE to STABLE

#define PULSE_MAX_VALID 2300
#define IDLE_TRESHOLD (PULSE_MAX_VALID + 500)  // Receiving is considered finished when IDLE_TRESHOLD us no edge change (ir) occurs.

#define PERIOD_MIN_VALID 25000                 // -> 40Hz min pwm frequency
#define PERIOD_MAX_VALID 12500                 // -> 80Hz max pwm frequency

#define AUTO_ZERO_CALIBRATION true             // default for auto zero calibration
#define ZERO_CALC_LOWER_LIMIT 20               // lower limit in %. current pulse in us - ZERO_CALC_LOWER_LIMIT
#define ZERO_CALC_UPPER_LIMIT 20               // upper limit in %. current pulse in us + ZERO_CALC_UPPER_LIMIT

#define SCALE_PULSE_MIN 1000                   // scaled pulse min (default value)
#define SCALE_PULSE_NEUTRAL 1500               // scaled pulse neutral (default value)
#define SCALE_PULSE_MAX 2000                   // scaled pulse max (default value)

#define AUTO_MIN_MAX_CALIBRATION true          // default for auto min/max calibration (default value)
#define MIN_MAX_CALIBRATE_LOWER_LIMIT 20       // lower limit in % based on calculated pulse_min
#define MIN_MAX_CALIBRATE_UPPER_LIMIT 20       // upper limit in % based on calculated pulse_max
/*
This means the distance between neutral and min/max.
In the standard, this is calculated from (max - min)/2.
The respective validity range is then calculated with:
Neutral +/- (VALID_MIN_MAX_MID +/- calibrate limits (%)).

Example No AutoZero and default values
minLowerLimit = (1500 - 500) - (1500 - 500) * 20%) = 800
maxLowerLimit = (1500 - 500) + (1500 - 500) * 20%) = 1200
minupperLimit = (1500 + 500) - (1500 + 500) * 20%) = 1600
maxLowerLimit = (1500 + 500) + (1500 + 500) * 20%) = 2400

Example AutoZero neutral=1300 and default values
minLowerLimit = (1300 - 500) - (1300 - 500) * 20%) = 640
maxLowerLimit = (1300 - 500) + (1300 - 500) * 20%) = 960
minupperLimit = (1300 + 500) - (1300 + 500) * 20%) = 1440
maxLowerLimit = (1300 + 500) + (1300 + 500) * 20%) = 2160
*/
#define VALID_MIN_MAX_MID ((SCALE_PULSE_MAX - SCALE_PULSE_MIN) / 2)

#define AUTO_MIN_MAX_CALIBRATE_TIME_SEC 5   // 5 sec for calculate min max
#define GET_FAILESAFE true                  // returns pulse_neutral if state not STABLE (default value)
#define GET_FAILESAFE_DURING_MIN_MAX false  // send failsafe value during auto_min_max_calibrate

// helpers
#define MIN(A, B) ((A) <= (B) ? (A) : (B))
#define MAX(A, B) ((A) >= (B) ? (A) : (B))

/*!
 * @brief states for SM
 */
enum class PwmState { NOT_CONNECTED,
                      UNSTABLE,
                      STABLE };

/*!
 * @brief Configuration of a channel
 */
struct pwm_config {
    uint16_t pulse_min;      ///< Minimum valid pulse width
    uint16_t pulse_max;      ///< Maximum valid pulse width
    uint16_t pulse_neutral;  ///< Neutral pulse width
    uint16_t period_max;     ///< Maximum period (fmin = 20Hz)(needed for FSM to detect NO_SIGNAL)
    boolean auto_zero;       ///< if true, auto calculate pulse_neutral if state change from NOT_CONNECTED to STABLE default is AUTO_ZERO_CALIBRATION
    boolean auto_min_max;    ///< if true, auto calculate pulse_min/max if state change from NOT_CONNECTED to STABLE default is AUTO_MIN_MAX_CALIBRATION
    boolean failsafe;        ///< if true and state not STABLE get impulse returns pulse_neutral default is GET_FAILESAFE
};

/*!
 * @brief Data for a PWM channel
 */
struct pwm_data {
    volatile int64_t lastIr = 0;               ///< Timestamp of last interrupt (IDLE_TRESHOLD after falling edge)
    volatile uint32_t highus = 0;              ///< Pulse width of the current pulse
    volatile uint32_t lastTotalus = 0;         ///< width of the last period in us (needed for FSM to detect jitter)
    volatile uint32_t totalus = 0;             ///< width of the current period in us
    volatile uint16_t frq = 0;                 ///< Frequency of the current pulse in Hz
    uint8_t stableCounter = 0;                 ///< Counter for detecting stability
    uint16_t auto_pulse_min = 0;               ///< detectet min pulse width in the first 5 sec after first STABLE
    uint16_t auto_pulse_max = 0;               ///< detectet max pulse width in the first 5 sec after first STABLE
    int64_t stable_time = 0;                   ///< Timestamp of last STABLE state
    PwmState state = PwmState::NOT_CONNECTED;  ///< State of the channel (NOT_CONNECTED, UNSTABLE, or STABLE)
};

/*!
 * @brief Object representing a PWM channel
 */
typedef struct pwm_channel {
    uint8_t pin;               ///< GPIO pin for the channel
    uint8_t channel;           ///< RMT channel for the channel
    struct pwm_config config;  ///< Configuration for the channel
    struct pwm_data data;      ///< Data for the channel
} pwm_channel_t;

/*!
 * @brief Initializes the PWM reader with the given channel pins
 * @param channelPins An array of GPIO pins for each channel
 * @param numberOfPins The number of channels to initialize
 */
void pwm_reader_init(const uint8_t channelPins[], const int numberOfPins);

/*!
 * @brief configure rmt_channels and start reading
 * @return
 *     - ESP_OK Success
 *     - ESP_ERR_INVALID_ARG Function pointer error.
 *     - ESP_FAIL System driver installed, can not register ISR handler for RMT
 */
esp_err_t pwm_reader_begin();

/*!
 * @brief Returns the object representing all PWM channels for debug
 * @return A pointer to an array of pwm_channel_t objects
 */
const pwm_channel_t *pwm_get_whole_channels_object();

/*!
 * @brief Returns the data for a specific PWM channel
 * @param channel The channel number
 * @return A pointer to the pwm_data object for the channel
 */
const pwm_data *pwm_get_channel_data(const uint8_t channel);

/*!
 * @brief Returns the configuration for a specific PWM channel
 * @param channel The channel number
 * @return A pointer to the pwm_config object for the channel
 */
const pwm_config *pwm_get_channel_config(const uint8_t channel);

/*!
 * @brief configure channel RMT_Channel
 * @param channel to configure
 * @param rmtChannel new RMT_Channel
 */
void pwm_set_channel_rmtChannel(const uint8_t channel, const uint8_t rmtChannel);


/*!
 * @brief configure channel pulse_min
 * @param channel to configure
 * @param pulse_min new config.pulse_min
 */
void pwm_set_channel_pulse_min(const uint8_t channel, const uint16_t pulse_min);

/*!
 * @brief configure channel pulse_max
 * @param channel to configure
 * @param pulse_max new config.pulse_min
 */
void pwm_set_channel_pulse_max(const uint8_t channel, const uint16_t pulse_max);

/*!
 * @brief configure channel pulse_neutral
 * @param channel to configure
 * @param pulse_neutral new config.pulse_min
 */
void pwm_set_channel_pulse_neutral(const uint8_t channel, const uint16_t pulse_neutral);

/*!
 * @brief configure channel period_max
 * @param channel to configure
 * @param pulse_min new config.period_max
 */
void pwm_set_channel_period_max(const uint8_t channel, const uint16_t period_max);

/*!
 * @brief configure channel auto_zero
 * @param channel to configure
 * @param pulse_min new config.auto_zero
 */
void pwm_set_auto_zero(const uint8_t channel, const boolean auto_zero);

/*!
 * @brief configure channel auto_min_max
 * @param channel to configure
 * @param pulse_min new config.auto_min_max
 */
void pwm_set_auto_min_max(const uint8_t channel, const boolean auto_min_max);

/*!
 * @brief configure channel failsafe
 * @param channel to configure
 * @param failsafe new config.failsafe
 */
void pwm_set_failsafe(const uint8_t channel, const boolean failsafe);

/*!
 * @brief set state e.g. for reset channel to new calibrate
 * @param channel to configure
 * @param state new state
 */
void pwm_set_state(const uint8_t channel, const PwmState state);

/*!
 * @brief Returns the last impuls width
 * @param channel The channel number
 * @return last impuls width for the channel in us
 *         if data->state != STABLE and config.failsafe==true -> pulse_neutral
 *         if data->state != STABLE and config.failsafe==false -> -1
 */
const int32_t pwm_get_rawPwm(const uint8_t channel);

/*!
 * @brief Returns the last scaled impuls width.  Linear scaling is performed.
 *        If the pulse duration <= pulse_neutral, scaling is performed between pulse_min and pulse_neutral.
 *        If the pulse duration > pulse_neutral, scaling is performed between pulse_neutral and pulse_max.
 * @param channel The channel number
 * @return last impuls width for the channel in us
 *         if data->state != STABLE and config.failsafe==true -> pulse_neutral
 *         if data->state != STABLE and config.failsafe==false -> -1
 */
const int32_t pwm_get_scaledPwm(const uint8_t channel);

/*!
 * @brief Get the offset of the channel to channel0
 *        If state of Channel or Channel0 != STABLE returns -1 or freq of Channel0 != freq of Channel
 *        If Channel = 0 returns 0
 * @param channel The channel number
 * @return offset in us between channel and channel0 or -1 if state != STABLE or 0 if channel == 0
 */
const int32_t pwm_get_offset(const uint8_t channel);

/*!
 * @brief Returns the actual state for a PWM channel
 * @param channel The channel number
 * @return The state of the channel (NOT_CONNECTED, UNSTABLE, or STABLE)
 */
const PwmState pwm_get_state(const uint8_t channel);

/*!
 * @brief Returns the actual state name for a PWM channel
 * @param channel The channel number
 * @return The state name of the channel (NOT_CONNECTED, UNSTABLE, or STABLE)
 */
const char *pwm_get_state_name(const uint8_t channel);

/*!
 * @brief Returns the actual frequency for a PWM channel
 * @param channel The channel number
 * @return The frequency of the channel
 */
const uint16_t pwm_get_frequency(const uint8_t channel);

/*!
 * @brief calculate duty cycle baset on pulse_min, pulse_max, pulse_current
 * @param pulse_min  Minimum valid pulse width
 * @param pulse_max  Maximum valid pulse width
 * @param pulse_current pulse width to calculate duty cycle
 * @return The calculated duty cycle
 */
const float_t pwm_get_rc_duty_cycle(const uint16_t pulse_min, const uint16_t pulse_max, const uint16_t pulse_current);

/*!
 * @brief deregister isr, delete pwm_channels, etc
 * @return some error or ESP_OK
 */
esp_err_t pwm_cleanup();

#endif /* esp32_rmt_pwm_reader_h */


src/esp32_pwm_read.cpp
/**
*********************************************************************************
* @file:	esp32_pwm_read.cpp
* @author:	R.W.
* @date:	2023.03.27
* @brief:	test program for use lib esp32_pwm_reader
*
* @history:	* 06.05.2023   rw v.0.1
*********************************************************************************
**/

#include <Arduino.h>
#include <Ticker.h>    // for wait in loop()
#include <inttypes.h>  // for printf

#include <esp32-rmt-pwm-reader.h>

Ticker ticker;

// init channels and pins
uint8_t pins[] = {13, 14, 33, 27, 32};  // Input pin numbers (pin 34 & 35 only usable as inputs!)
int numberOfChannels = sizeof(pins) / sizeof(uint8_t);

// helper for table print
#define CLEAR_SCREEN printf("\e[1;1H\e[2J")

void gotoRowCol(const int row, const int col) {
    // Position the cursor at the desired position (row, col)
    Serial.print("\033[");  // Begin of escape sequence
    Serial.print(row);      // row number (begins with 1)
    Serial.print(";");
    Serial.print(col);      // column (begins with 1)
    Serial.print("H");
}


//
// =======================================================================================================
// READ PWM RC SIGNALS AND PRINT TABLE
// =======================================================================================================
//
void readPwmSignals() {
    // auto channels = pwm_get_whole_channels_object(); // test read the whole array of channels readonly

    gotoRowCol(1, 1);
    
    // print table
    /*
    The function uses tab escape secence to format the table.
    It assumes that the terminal window has a tab width of 8. 
    If this is not the case, the tab width of the terminal shell should be set to 8 outside the application.
    Depending on the shell used, there are individual methods.
    Maybe this link can be helpful:
    https://stackoverflow.com/questions/10782699/how-to-set-4-space-tab-in-bash
    */
    // table header
    printf("Channel\toffset\t    raw\t scaled\t period\tfreq Hz\t  p_min\t p-ntrl\t  p_max\t   dc %%\t  state\n");
    for (uint8_t channel = 0; channel < numberOfChannels; channel++) {
        auto data = pwm_get_channel_data(channel);      // the whole date struct for the channel
        auto config = pwm_get_channel_config(channel);  // the whole config struct for the channel

       
        /* // If you want only the channels with status=STABLE
        if (data->state != PwmState::STABLE)
            continue;
        */

        auto rawPwm = pwm_get_rawPwm(channel);  // if you use pwm_get_channel_data() you can use date->higus
        auto scaledPwm = pwm_get_scaledPwm(channel);
    
        Serial.printf("%7d\t", channel);
        Serial.printf("%7" PRId32 "\t", pwm_get_offset(channel) );
        Serial.printf("%7lu\t", rawPwm);
        Serial.printf("%7lu\t", scaledPwm);
        Serial.printf("%7lu\t", data->totalus);
        Serial.printf("%7lu\t", data->frq);
        Serial.printf("%7u\t", config->pulse_min);
        Serial.printf("%7u\t", config->pulse_neutral);
        Serial.printf("%7u\t", config->pulse_max);
        Serial.printf("%7.2f\t", pwm_get_rc_duty_cycle(config->pulse_min,
                                                       config->pulse_max,
                                                       (uint16_t)rawPwm));
        Serial.printf("%7lu\t", data->state);
        Serial.printf(" %s\n", pwm_get_state_name(channel));
    }
}

void setup() {
    Serial.begin(115200);
    Serial.println("V0.03 280623 0101");

    // begin prepare pwm reading -------------------------------------------------
    // init channels
    pwm_reader_init(pins, numberOfChannels);

    // here you can change channel defaults values before reading (if needed)
    // e.g. pwm_set_channel_pulse_min() /max/neutral
    // e.g. set auto_zero/auto_min_max for channel 0-2
    for (int ch = 0; ch < 3; ch++) {
        pwm_set_auto_zero(ch, true);     // set channel to auto zero
        pwm_set_auto_min_max(ch, true);  // set channel to auto min/max calibration
    }

    // begin reading 
    esp_err_t err = pwm_reader_begin();
    if (err != ESP_OK) {
        Serial.printf("begin() err: %i", err);
    }
    // end prepare pwm reading ----------------------------------------------------

    // DEBUG:
    // if you want to read the current configuration, you can use the function pwm_get_channel_config(x)
    auto cfg = pwm_get_channel_config(0);  // read conbfig for channel 0
    ESP_LOGV(TAG, "CH: 0 min:%u max:%u neutral:%u periodMax:%u autoZero:%d",
             cfg->pulse_min,
             cfg->pulse_max,
             cfg->pulse_neutral,
             cfg->period_max,
             cfg->auto_zero);

    CLEAR_SCREEN;  // clear screen before table print
    
    ticker.attach_ms(100, readPwmSignals);  // print table every 100ms
}

void loop() {
}
